text/0000-partial_types.md

```diff
@@ -0,0 +1,672 @@
+    p1 : &mut %{self.{x, y, z, w}} Point;
+```
+
+Where :
+ - `.{<fld1>, <fld2>, }` is an field-set quasi-field
+
+We assume, that each field could be in one of 2 specific field-integrity - `%fit` and `%deny`. 
+
+We also must reserve as a keyword a `%miss` field-integrity for future ReExtendeded Partial Types, which allows to create **safe** self-referential types.
+
+`%fit` is default field-integrity and it means we have an access to this field and could use it as we wish. But if we try to access `%deny` field it cause a compiler error.
+
+```rust
+// case (C3)
+let foo = 0i16;
+    foo : %{self.*} i16  ~  %{self.self} i16;
+    foo : %{%fit self.*} i16  ~  %{%fit self.self} i16;
+    foo : %{%fit self.*, %deny self._} i16;
+
+
+// case (C4)
+// FROM case (C2)
+let &mut p1 : &mut Point = Point {x:1.0, y:2.0, z:3.0, t:4.0, w:5.0};
+    p1 : &mut %{%fit self.*} Point  ~  &mut  %{%fit self.*, %deny self._};
+    p1 : &mut %{%fit self.{x, y, z, w}} Point;
+```
+
+Where :
+ - `%fit` integrity
+ - `%deny` integrity
+ - `._` is a "rest of fields" quasi-field
+
+As we see, 
+ - `%empty : %{%deny self.*}` or `%empty : %{}` integrity
+ - `%full  : %{%fit  self.*}` or `%full  : %{self.*}` integrity
+
+
+## Partial parameters
+
+We add enough integrity, and could write partial parameters for non-virtual function declarations:
+```rust
+// case (D1)
+fn re_ref_t (& p : & %{self.t, %ignore self._} Point) -> &f64 {
+   &p.t
+}
+
+// case (D2)
+fn refmut_w (&mut p : &mut %{self.w, %ignore self._} Point) -> &mut f64 {
+   &mut p.w
+}
+```
+
+Where :
+ - `%ignore` is a "don't care which exactly" quasi filed-integrity (`%_` is a whole type integrity and it is unclear if we could use it in both contents)
+
+But `%ignore self._` quasi-filed-integrity of quasi-field looks annoying, so we simplify a bit adding `%any : %ignore self._`.
+
+```rust
+// case (D3)
+// FROM case (D1)
+fn re_ref_t (& p : & %{self.t, %any} Point) -> &f64 {
+   &p.t
+}
+
+// case (D4)
+// FROM case (D2)
+fn refmut_w (&mut p : &mut %{self.w, %any} Point) -> &mut f64 {
+   &mut p.w
+}
+
+// case (D5)
+struct PointExtra {
+    x: f64,
+    y: f64,
+    saved_x: f64,
+    saved_y: f64,
+}
+
+fn x_store(&mut p1 : &mut %{self.saved_x, %any} PointExtra, & p2 : & %{self.x, %any} PointExtra) {
+    *p1.saved_x = *p2.x
+}
+
+fn x_restore(&mut p1 : &mut %{self.x, %any} PointExtra, & p2 : & %{self.saved_x, %any} PointExtra) {
+    *p1.x = *p2.saved_x;
+}
+```
+
+or use `where` clause if integrity is extra verbose:
+```rust
+// case (D6)
+// FROM case (D5)
+
+fn x_store(&mut p1 : &mut %fit_sv_x PointExtra, & p2 : & %fit_x PointExtra) 
+    where %fit_sv_x : %{self.saved_x, %any},
+          %fit_x : %{self.x, %any}
+{
+    *p1.saved_x = *p2.x
+}
+
+fn x_restore(&mut p1 : &mut %fit_x PointExtra, & p2 : & %fit_sv_x PointExtra) 
+    where %fit_sv_x : %{self.saved_x, %any},
+          %fit_x : %{self.x, %any}
+{
+    *p1.x = *p2.saved_x;
+}
+```
```

Comment:
This doesn't seem like good syntax to me, since where is usually used to indicate bounds on parameters, whereas this is defining parameters instead. Wouldn't it make more sense to simply use type aliases for this?

For example, `type PointSaveXMut<'a> = &'a mut %{self.saved_x, %any} PointExtra` could then allow using `PointSaveXMut<'_>` directly.

Reply:
Good point! I like it!